### PR TITLE
Better GC heuristics and ergonomics control.

### PR DESCRIPTION
**Pull request by olonho**

### CODE REVIEW

File: runtime/src/main/cpp/Memory.cpp [hunk 1]
```diff
@@ -1419,7 +1431,7 @@ void garbageCollect(MemoryState* state, bool force) {
   decrementStack(state);
   size_t afterDecrements = state->toRelease->size();
   ssize_t stackReferences = afterDecrements - beforeDecrements;
-  if (stackReferences * 5 > state->gcThreshold) {
+  if (state->gcErgonomics && stackReferences * 5 > state->gcThreshold) {
```
Comment (LepilkinaElena) [inline runtime/src/main/cpp/Memory.cpp]: Compilation error in case if GC_ERGONOMICS is 0, isn't it?

File: runtime/src/main/cpp/Memory.cpp [hunk 2]
```diff
@@ -1917,17 +1945,39 @@ void startGC() {
...
+}
+
+KLong getGCThresholdAllocations() {
+  GC_LOG("getGCThresholdAllocation\n")
+  return memoryState->allocSinceLastGcThreshold;
+}
+
+void setTuneGCThreshold(KBoolean value) {
+  GC_LOG("setTuneGCThreshold %d\n", value)
+  memoryState->gcErgonomics = value;
```
Comment (LepilkinaElena) [inline runtime/src/main/cpp/Memory.cpp]: Compilation error in case if GC_ERGONOMICS is 0

File: runtime/src/main/cpp/Memory.cpp [hunk 3]
```diff
@@ -1917,17 +1945,39 @@ void startGC() {
...
+}
+
+KBoolean getTuneGCThreshold() {
+  GC_LOG("getTuneGCThreshold %d\n")
+  return memoryState->gcErgonomics;
```
Comment (LepilkinaElena) [inline runtime/src/main/cpp/Memory.cpp]: Same here

File: runtime/src/main/kotlin/kotlin/native/internal/GC.kt [hunk 4]
```diff
@@ -63,9 +63,36 @@ object GC {
         get() = getThreshold()
         set(value) = setThreshold(value)
 
+    /**
+     * GC allocation threshold, controlling how many bytes allocated since last
+     * collection will trigger new GC.
+     */
+    var thresholdAllocations: Long
+        get() = getThresholdAllocations()
+        set(value) = setThresholdAllocations(value)
```
Comment (LepilkinaElena) [inline runtime/src/main/kotlin/kotlin/native/internal/GC.kt]: Not used now as I understood, value isn't changed. Is it for future tuning?

Comment (olonho) [reply runtime/src/main/kotlin/kotlin/native/internal/GC.kt]: It's user API for GC tuning.

**Pull request merged**